st2actions/st2actions/workflows/workflows.py

```diff
@@ -85,6 +97,76 @@ def process(self, message):
             # the database and fail the workflow execution gracefully. In this case,
             # the garbage collector will find and cancel these workflow executions.
             self.fail_workflow_execution(message, e)
+        finally:
+            if type(message) == wf_db_models.WorkflowExecutionDB:
+                self._handling_workflows.discard(str(message["action_execution"]))
+            else:
+                try:
+                    self._handling_tasks.discard(
+                        "__".join([str(message.id), str(message["parent"])])
+                    )
+                except ValueError:
+                    pass
+
+    def shutdown(self):
+        super(WorkflowExecutionHandler, self).shutdown()
+        self._shutdown = True
+        if self._helper_thread:
+            concurrency.kill(self._helper_thread)
+
+        if cfg.CONF.workflow_engine.graceful_shutdown:
```

Comment:
This is not resolved yet. We don't need a `graceful_shutdown` config option. The workflow engine should alway graceful shutdown.